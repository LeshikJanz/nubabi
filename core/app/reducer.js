// @flow
import type { Action, AppState } from '../types';

export const initialState: AppState = {
  started: false,
  online: false,
  error: null,
  success: null,
};

const reducer = (state: AppState = initialState, action: Action): AppState => {
  // Map all app errors into state.app.error.
  // In React Native, we show errors in one nicely animated unobtrusive alert.
  // In the browser, we prefer local error messages rendering.
  // TODO: Refactor it. We don't want sticky strings.
  if (action.type.endsWith('_FAILURE')) {
    // $FlowFixMe
    state = { ...state, error: action.payload }; // eslint-disable-line no-param-reassign
  }

  switch (action.type) {
    case 'APP_ONLINE':
      return { ...state, error: null, online: action.payload.online };

    case 'APP_STARTED':
      return { ...state, started: true };

    case 'APP_ERROR':
      return { ...state, error: action.payload };

    case 'APP_SUCCESS': {
      return { ...state, error: null, success: action.payload };
    }

    case 'APP_INFO': {
      return { ...state, error: null, info: action.payload };
    }

    case 'ALERT_SHOWN': {
      return { ...state, error: null, success: null };
    }

    default:
      return state;
  }
};

export default reducer;
