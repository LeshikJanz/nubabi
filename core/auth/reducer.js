// @flow
import type { Action, AuthState } from '../types';
import { merge } from 'ramda';

export const initialState = {
  isAuthenticated: false,
  isFetching: false,
  token: null,
  errorMessage: '',
};

const reducer = (
  state: AuthState = initialState,
  action: Action,
): AuthState => {
  switch (action.type) {
    case 'LOGIN_REQUEST':
      return merge(state, {
        isFetching: true,
        isAuthenticated: false,
      });
    case 'LOGIN_SUCCESS':
      return merge(state, {
        isFetching: false,
        isAuthenticated: true,
      });
    case 'LOGIN_FAILURE':
      return merge(state, {
        isFetching: false,
        isAuthenticated: false,
        errorMessage: action.payload.message,
      });
    case 'ON_AUTH': {
      return merge(state, {
        isFetching: false,
        isAuthenticated: !!action.payload.user,
        errorMessage: null,
        token: action.payload.token,
      });
    }
    case 'LINK_ACCOUNT_REQUEST': {
      return merge(state, {
        isFetching: true,
      });
    }
    case 'LINK_ACCOUNT_SUCCESS': {
      return merge(state, {
        isFetching: false,
      });
    }
    case 'LINK_ACCOUNT_FAILURE': {
      return merge(state, {
        isFetching: false,
      });
    }
    case 'UNLINK_ACCOUNT_REQUEST': {
      return merge(state, {
        isFetching: true,
      });
    }
    case 'UNLINK_ACCOUNT_SUCCESS': {
      return merge(state, {
        isFetching: false,
      });
    }
    case 'UNLINK_ACCOUNT_FAILURE': {
      return merge(state, {
        isFetching: false,
      });
    }
    case 'LOGOUT':
      return initialState;
    default:
      return state;
  }
};

export default reducer;
