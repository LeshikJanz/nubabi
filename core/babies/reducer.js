// @flow
import type { Action, BabyState } from '../types';
import { path } from 'ramda';

export const initialState = {
  isFetching: false,
  failure: false,
  currentBabyId: null,
  currentBabyPhoto: null,
};

const reducer = (
  state: BabyState = initialState,
  action: Action,
): BabyState => {
  switch (action.type) {
    case 'SET_BABY_PHOTO':
      return Object.assign({}, state, {
        currentBabyPhoto: action.payload,
      });
    case 'GET_BABIES_REQUEST':
      return Object.assign({}, state, {
        isFetching: true,
      });
    case 'GET_BABIES_SUCCESS':
      return Object.assign({}, state, {
        isFetching: false,
        failure: false,
        currentBabyId:
          state.currentBabyId || path(['payload', '0', 'id'], action),
      });
    case 'GET_BABIES_FAILURE': {
      return Object.assign({}, state, {
        isFetching: false,
        failure: true,
      });
    }
    case 'SELECT_BABY': {
      return {
        ...state,
        currentBabyId: action.payload,
      };
    }
    default:
      return state;
  }
};

export default reducer;
