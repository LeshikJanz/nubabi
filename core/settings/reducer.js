// @flow
import type { SettingsSetValueAction, SettingsState, State } from '../types';
import { assocPath, merge } from 'ramda';

type SettingsAction = SettingsSetValueAction;

export const initialState: SettingsState = {
  unitDisplay: {
    weight: 'kg',
    height: 'cm',
  },
  notifications: {
    memories: true,
    stimulation: true,
    growth: true,
    activities: false,
    email: true,
  },
  memories: {
    displaySuggestions: true,
  },
};

export function setSettingsValue(path: Array<string>, value: any) {
  return {
    type: 'SETTINGS_SET_VALUE',
    payload: {
      path,
      value,
    },
  };
}

export const resetSettings = () => ({
  type: 'RESET_SETTINGS',
});

export const unitDisplaySelector = (state: State) => state.settings.unitDisplay;

export default function reducer(
  state: SettingsState = initialState,
  action: SettingsAction,
) {
  switch (action.type) {
    case 'SETTINGS_SET_VALUE': {
      return assocPath(action.payload.path, action.payload.value, state);
    }
    case 'RESET_SETTINGS': {
      return initialState;
    }
    default: {
      return state;
    }
  }
}
