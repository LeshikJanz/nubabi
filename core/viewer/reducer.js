// @flow
import type { ViewerState, Action } from '../types';

export const initialState = null;

const reducer = (state: ?ViewerState = initialState, action: Action) => {
  switch (action.type) {
    case 'ON_AUTH': {
      return action.payload.user;
    }

    default: {
      return state;
    }
  }
};

export default reducer;
