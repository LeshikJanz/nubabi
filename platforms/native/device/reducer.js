// @flow
import type { MobileDeviceState, DeviceState, Action } from 'core/types';
import { assoc } from 'ramda';

const ORIENTATION_CHANGED = 'device/ORIENTATION_CHANGED';

const initialState = {
  isReactNative: true,
  deviceModel: null,
  deviceId: null,
  appVersion: null,
  systemVersion: null,
  bundleId: null,
  locale: null,
  orientation: null,
};

export const onOrientationChange = (orientation: 'LANDSCAPE' | 'PORTRAIT') => ({
  type: 'device/ORIENTATION_CHANGED',
  payload: orientation,
});

/* eslint-disable no-unused-vars */
const reducer = (
  state: MobileDeviceState = initialState,
  action: Action,
): DeviceState => {
  switch (action.type) {
    case ORIENTATION_CHANGED: {
      if (state.orientation !== action.payload) {
        return assoc('orientation', action.payload, state);
      }
    }
    default: {
      return state;
    }
  }
};

/* eslint-enable no-unused-vars */
export default reducer;
