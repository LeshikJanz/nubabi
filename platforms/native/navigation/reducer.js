// @flow
import type { State, Action } from '../../../core/types';
import AppNavigator from './AppNavigator';

const reducer = (state: State, action: Action) => {
  const newState = AppNavigator.router.getStateForAction(action, state);
  return newState || state;
};

export default reducer;
