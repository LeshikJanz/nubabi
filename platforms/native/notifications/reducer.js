// @flow
import { Observable } from 'rxjs/Observable';
import firebase from 'react-native-firebase';

import type { Action } from 'core/types';

const notificationReceivedEpic = action$ => {
  const onMessage$ = Observable.create(observer => {
    const unsubscribe = firebase.messaging().onMessage(message => {
      observer.next({ type: 'NOTIFICATION_RECEIVED', payload: message });
    });

    return unsubscribe;
  });

  const streams: Array<any> = [onMessage$];

  return action$
    .filter((action: Action) => action.type === 'APP_STARTED')
    .mergeMap(() => Observable.merge(...streams));
};

export const epics = [notificationReceivedEpic];
