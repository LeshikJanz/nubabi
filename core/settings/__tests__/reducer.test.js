import reducer, {
  initialState, setSettingsValue, resetSettings,
  unitDisplaySelector,
} from '../reducer';

describe('settings reducer', () => {
  it('has an initial state', () => {
    expect(reducer(undefined, { type: '@@INIT' })).toMatchSnapshot();
  });

  it('handles RESET_SETTINGS', () => {
    expect(reducer({
      ...initialState,
      unitDisplay: {
        weight: 'lb',
        height: 'cm',
      }
    }, resetSettings())).toEqual(initialState);
  });

  it('handles SETTINGS_SET_VALUE', () => {
    expect(
      reducer(initialState, setSettingsValue(['notifications', 'memories'])),
    ).toMatchSnapshot();
  });
});

describe('settings actions', () => {
  describe('resetSettings', () => {
    it('creates a RESET_SETTINGS action', () => {
      expect(resetSettings()).toMatchSnapshot();
    });
  });
});

describe('unitDisplaySelector', () => {
  it('returns the unit display preferences', () => {
    const state = { settings: initialState };

    expect(unitDisplaySelector(state)).toMatchObject({
      weight: 'kg',
      height: 'cm'
    });
  })
});
