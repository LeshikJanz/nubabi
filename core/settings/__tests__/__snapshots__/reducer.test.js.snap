// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings actions resetSettings creates a RESET_SETTINGS action 1`] = `
Object {
  "type": "RESET_SETTINGS",
}
`;

exports[`settings reducer handles SETTINGS_SET_VALUE 1`] = `
Object {
  "memories": Object {
    "displaySuggestions": true,
  },
  "notifications": Object {
    "activities": false,
    "email": true,
    "growth": true,
    "memories": undefined,
    "stimulation": true,
  },
  "unitDisplay": Object {
    "height": "cm",
    "weight": "kg",
  },
}
`;

exports[`settings reducer has an initial state 1`] = `
Object {
  "memories": Object {
    "displaySuggestions": true,
  },
  "notifications": Object {
    "activities": false,
    "email": true,
    "growth": true,
    "memories": true,
    "stimulation": true,
  },
  "unitDisplay": Object {
    "height": "cm",
    "weight": "kg",
  },
}
`;
